File: server/nginx/app/todo/api/dd/task.lua

local dd_task    = _load "$dd_task"
local utils      = _load "%utils"
local hashid     = _load "#hashid"

local __ = { ver = "v22.12.07" }

__.get__ = {
    "获取待办任务",
    req = {
        { "user_id", "用户编码" },
        { "task_id", "任务编码" },
    },
    res = "$dd_task"
}
__.get = function(t)

    local  res, err = dd_task.get { user_id = t.user_id, task_id = t.task_id }
    if not res then return nil, err and "服务器忙，请稍候再试" or "待办任务不存在" end

    return res

end

__.list__ = {
    "任务列表",
    req = {
        { "user_id"      , "用户编码" },
        { "task_cate_id?", "列表编码" },
    },
    res = "$dd_task[]"
}
__.list = function(t)

    local wh = {
        user_id      = t.user_id,
        task_cate_id = t.task_cate_id or ""
    }

    local res = dd_task.list(wh)
    if not res then return nil, "服务器忙，请稍候再试" end

    return res

end

__.add__ = {
    "添加待办任务",
    req = {
        { "user_id"      , "用户编码" },
        { "task_cate_id?", "列表编码" },
        { "task_name"    , "任务名称" },
    },
    res = "$dd_task"
}
__.add = function(t)

    -- @t : $dd_task
    t.task_id      = hashid.generate()
    t.task_cate_id = t.task_cate_id or ""
    t.task_name    = utils.strip_name(t.task_name)
    t.is_finished  = t.is_finished  or 0
    t.is_important = t.is_important or 0
    t.is_today     = t.is_today     or 0
    t.list_index   = t.list_index   or -1
    t.create_time  = ngx.localtime()
    t.update_time  = ngx.localtime()

    local  ok, err = dd_task.add (t)
    if not ok then return nil, err end

    return __.get(t)

end

__.set__ = {
    "修改待办任务",
    log = true,
    req = {
        { "user_id"       , "用户编码"            },
        { "task_id"       , "待办任务编码"        },
        { "task_name?"    , "任务名称"            },
        { "task_desc?"    , "任务描述"            },
        { "is_finished?"  , "是否完成", "number"  },
        { "is_important?" , "是否重要", "number"  },
        { "is_today?"     , "我的一天", "number"  },
        { "closing_date?" , "截止日期"            },
        { "list_index?"   , "任务排序", "number"  },
    },
    res = "$dd_task"
}
__.set = function(t)

    -- @t : $dd_task
    t.task_name = utils.strip_name(t.task_name)

    local  tOld, err = __.get(t)
    if not tOld then return nil, err end

    -- 生成待更新的数据
    -- @d : $dd_task
    local  d, wh = utils.gen_update(t, tOld, "user_id", "task_id" )
    if not d then return tOld end

    t.update_time = ngx.localtime()

    local  ok, err = dd_task.set(wh)
    if not ok then return nil, err end

    return __.get(t)

end

__.set_name__ = {
    "修改任务名称",
    req = {
        { "user_id"   , "用户编码"  },
        { "task_id"   , "任务编码"  },
        { "task_name" , "任务名称"  },
    },
    res = "$dd_task"
}
__.set_name = function(t)
    t.task_name = utils.strip_name(t.task_name)
    return __.set(t)
end

__.set_is_finished__ = {
    "设置是否重要",
    log = true,
    req = {
        { "user_id"     , "用户编码"            },
        { "task_id"     , "任务编码"            },
        { "is_finished" , "是否完成", "number"  },
    },
    res = "$dd_task"
}
__.set_is_finished = function(t)
    return __.set(t)
end

__.set_is_important__ = {
    "设置是否重要",
    req = {
        { "user_id"     , "用户编码"            },
        { "task_id"     , "任务编码"            },
        { "is_important", "是否重要", "number"  },
    },
    res = "$dd_task"
}
__.set_is_important = function(t)
    return __.set(t)
end

__.set_myday__ = {
    "设置为我的一天",
    log = true,
    req = {
        { "user_id" , "用户编码" },
        { "task_id" , "任务编码" },
        { "myday"   , "我的一天" },
    },
    res = "$dd_task"
}
__.set_myday = function(t)
    return __.set(t)
end

__.set_closing_date__ = {
    "设置截止日期",
    log = true,
    req = {
        { "user_id"      , "用户编码" },
        { "task_id"      , "任务编码" },
        { "closing_date" , "截止日期" },
    },
    res = "$dd_task"
}
__.set_closing_date = function(t)

    local  tOld, err = __.get(t)
    if not tOld then return nil, err end

    -- 生成待更新的数据
    -- @d : $dd_task
    local  d, wh = utils.gen_update(t, tOld, "user_id", "task_id" )
    if not d then return tOld end

    t.update_time = ngx.localtime()

    local  ok, err = dd_task.set(wh)
    if not ok then return nil, err end

    return __.get(t)

end

__.del__ = {
    "删除任务",
    req = {
        { "user_id" , "用户编码" },
        { "task_id" , "任务编码" },
    },
    res = "boolean"
}
__.del = function(t)

    local  ok, err = dd_task.del { user_id = t.user_id, task_id = t.task_id }
    if not ok then return nil, err end

    return true

end

return __

File: todolist/src/store/modules/task.ts

import { defineStore      } from 'pinia'
import { useAppStore      } from './app'
import { useTaskCateStore } from './task-cate'
import { router           } from '@/router'
import { reactive, toRefs, computed, markRaw } from 'vue'

const DEFAULT_DATE = '1900-01-01'
export const useTaskStore = defineStore('task', () => {
    const G        = window.G || {}
    const appStore = useAppStore()

    // 数据
    const m = reactive({
        serach_focus    : false                           , // 搜索框获得焦点
        search_val      : ''                              , // 搜索值
        tasks           : G.tasks || [] as $api.$dd_task[], // 待办任务列表
        curr_edit_item  : null as null | $api.$dd_task    , // 当前编辑项
        contextmenu_ref : null as any                     , // 右键菜单引用
    })

    // 任务分类
    const task$ = computed((): Record<string, $api.$dd_task[]> => {
        const myday_list        = [] as $api.$dd_task[] // 我的一天
        const important_list    = [] as $api.$dd_task[] // 重要
        const closing_date_list = [] as $api.$dd_task[] // 计划内
        const inbox_list        = [] as $api.$dd_task[] // 任务
        const custom_cate_list  = {} as Record<string,  $api.$dd_task[]>

        m.tasks.forEach(item => {
            // 我的一天
            if (item.myday === appStore.today$) myday_list.push(item)

            // 重要
            if (item.is_important === 1) important_list.push(item)

            // 计划内
            if (item.closing_date > DEFAULT_DATE) closing_date_list.push(item)

            // 自定义列表
            if (item.task_cate_id) {
                custom_cate_list[item.task_cate_id] = custom_cate_list[item.task_cate_id] || []
                custom_cate_list[item.task_cate_id].push(item)
            } else {
                inbox_list.push(item)
            }
        })

        return {
            myday       : myday_list,
            important   : important_list,
            closing_date: closing_date_list,
            inbox       : inbox_list,
            ...custom_cate_list
        }
    })

    // ---------------------------------------------------------------------------

    // 添加任务
    async function addTask(task_name: string) {
        const id      = (router.currentRoute.value.params.id || '') as string
        const cate_id = appStore.nav_ids.includes(id) ? '' : id
        const res = await $api.dd.task.add({ task_name: task_name, task_cate_id: cate_id }, { showLoading: true })
        if (res.ok) {
            m.tasks.push(res.data)
        }
        return res
    }

    // 修改任务名称
    async function setTaskName(item: $api.$dd_task) {
        const input_value = await $utils.showPrompt('请输入任务名称', '修改任务', {
            inputValue: item.task_name,
            inputValidator(value) { // value 默认为 null
                return (value && value.trim()) ? true : '任务名称不能为空'
            }
        })
        if ( !input_value ) return

        const res = await $api.dd.task.set_name({ task_id: item.task_id, task_name: input_value as string }, { showLoading: true })
        if ( !res.ok ) return

        const idx = m.tasks.findIndex(_ => _.task_id === item.task_id)
        if (idx !== -1) m.tasks.splice(idx, 1, res.data)
    }

    // 设置任务是否已完成标识
    async function setTaskIsFinished(item: $api.$dd_task) {
        const task_id     = item.task_id
        const is_finished = item.is_finished === 1 ? 0 : 1
        const res = await $api.dd.task.set_is_finished({ task_id, is_finished }, { showLoading: true })
        if (res.ok) {
            $utils.arr.replace(m.tasks, res.data, 'task_id')
        }
        return res
    }

    // 设置任务是否重要标识
    async function setTaskImportant(item: $api.$dd_task) {
        const task_id      = item.task_id
        const is_important = item.is_important === 1 ? 0 : 1
        const res = await $api.dd.task.set_is_important({ task_id, is_important })
        if (res.ok) {
            $utils.arr.replace(m.tasks, res.data, 'task_id')
        }
        return res
    }

    // 设置任务我的一天标识
    async function setTaskMyday(item: $api.$dd_task) {
        const task_id = item.task_id
        const myday   = item.myday !== DEFAULT_DATE ? DEFAULT_DATE : appStore.server_date
        const res = await $api.dd.task.set_myday({ task_id, myday })
        if (res.ok) {
            $utils.arr.replace(m.tasks, res.data, 'task_id')
        }
        return res
    }

    // 设置任务截止日期
    async function setTaskClosingDate(item: $api.$dd_task, closing_date: string = '') {
        if (closing_date === 'today'   ) closing_date = appStore.server_date
        if (closing_date === 'tomorrow') closing_date = $utils.dt.nextDate(appStore.server_time)
        if (closing_date === 'delete'  ) closing_date = DEFAULT_DATE

        const res = await $api.dd.task.set_closing_date({ task_id: item.task_id, closing_date })
        if (res.ok) {
            $utils.arr.replace(m.tasks, res.data, 'task_id')
        }
        return res
    }

    // 删除任务
    async function delTask(item: $api.$dd_task) {
        const confirm = await $utils.showConfirm('是否删除当前任务')
        if ( !confirm ) return

        const res = await $api.dd.task.del({ task_id: item.task_id })
        if ( !res.ok ) return

        const idx = m.tasks.findIndex(_ => _.task_id === item.task_id)
        if (idx !== -1) m.tasks.splice(idx, 1)
    }

    // ---------------------------------------------------------------------------

    // 设置搜索是否获得焦点
    function setSearchFocus(focus: boolean) {
        m.serach_focus = focus
    }

     // 退出清空数据
     function clear() {
        m.serach_focus = false
        m.search_val   = ''
        m.tasks        = []
    }

    // 设置当前右键编辑项
    function setCurrEditItem(item: $api.$dd_task | null) {
        useTaskCateStore().hideContextmenu()
        m.curr_edit_item = item
    }

    // 设置右键菜单引用
    function setContextmenuRef(ref: any) {
        m.contextmenu_ref = markRaw(ref) // 标记不进行 proxy 代理
    }

    // 隐藏右键菜单
    function hideContextmenu() {
        m.contextmenu_ref && m.contextmenu_ref.hide()
    }


    return {
        ...toRefs(m),
        task$,
        clear,
        setSearchFocus,
        setCurrEditItem,
        setContextmenuRef,
        hideContextmenu,
        // ----------------------
        addTask,
        delTask,
        setTaskName,
        setTaskIsFinished,
        setTaskImportant,
        setTaskMyday,
        setTaskClosingDate,
    }
})
